Observed a panic: "index out of range"

/kind bug

Version: canary

**What happened**:
Panic when the controller attempts to read an invalid/misconfigured `Certificate`
- Created everything but made a mistake when creating the Certificate and it didn't have any dnsNames or commonName
```
I1101 06:48:47.447567       1 server.go:68] Listening on http://0.0.0.0:9402
I1101 06:48:47.449045       1 leaderelection.go:174] attempting to acquire leader lease...
I1101 06:48:47.470629       1 leaderelection.go:184] successfully acquired lease kube-system/cert-manager-controller
I1101 06:48:47.470923       1 factory.go:58] NS does not exist on lookup for  - v1.GroupVersionKind{Group:"certmanager.k8s.io", Version:"v1alpha1", Kind:"Certificate"}
I1101 06:48:47.470985       1 factory.go:66] existing informer does not exist on lookup for  - v1.GroupVersionKind{Group:"certmanager.k8s.io", Version:"v1alpha1", Kind:"Issuer"}
I1101 06:48:47.471002       1 factory.go:66] existing informer does not exist on lookup for  - v1.GroupVersionKind{Group:"certmanager.k8s.io", Version:"v1alpha1", Kind:"ClusterIssuer"}
I1101 06:48:47.471017       1 factory.go:66] existing informer does not exist on lookup for  - v1.GroupVersionKind{Group:"", Version:"v1", Kind:"Secret"}
I1101 06:48:47.471030       1 factory.go:66] existing informer does not exist on lookup for  - v1.GroupVersionKind{Group:"extensions", Version:"v1beta1", Kind:"Ingress"}
I1101 06:48:47.471096       1 factory.go:70] existing informer found for  - v1.GroupVersionKind{Group:"certmanager.k8s.io", Version:"v1alpha1", Kind:"ClusterIssuer"}
I1101 06:48:47.471116       1 factory.go:58] NS does not exist on lookup for kube-system - v1.GroupVersionKind{Group:"", Version:"v1", Kind:"Secret"}
I1101 06:48:47.471153       1 factory.go:70] existing informer found for  - v1.GroupVersionKind{Group:"certmanager.k8s.io", Version:"v1alpha1", Kind:"Issuer"}
I1101 06:48:47.471169       1 factory.go:70] existing informer found for  - v1.GroupVersionKind{Group:"", Version:"v1", Kind:"Secret"}
I1101 06:48:47.572130       1 controller.go:140] clusterissuers controller: syncing item 'cert-manager-master-ci'
I1101 06:48:47.572200       1 factory.go:70] existing informer found for kube-system - v1.GroupVersionKind{Group:"", Version:"v1", Kind:"Secret"}
I1101 06:48:47.572287       1 controller.go:140] clusterissuers controller: syncing item 'cert-manager-master-ci-tui'
I1101 06:48:47.572334       1 factory.go:70] existing informer found for kube-system - v1.GroupVersionKind{Group:"", Version:"v1", Kind:"Secret"}
I1101 06:48:47.572553       1 controller.go:198] certificates controller: syncing item 'default/neo4j-cluster-ci-tls'
I1101 06:48:47.572604       1 factory.go:70] existing informer found for  - v1.GroupVersionKind{Group:"", Version:"v1", Kind:"Secret"}
I1101 06:48:47.572634       1 sync.go:98] Error checking existing TLS certificate: error decoding cert PEM block
I1101 06:48:47.572640       1 controller.go:198] certificates controller: syncing item 'kube-system/k8s-ui-init-ci-tls'
I1101 06:48:47.572668       1 sync.go:199] Preparing certificate with issuer
I1101 06:48:47.572695       1 factory.go:70] existing informer found for kube-system - v1.GroupVersionKind{Group:"", Version:"v1", Kind:"Secret"}
I1101 06:48:47.572718       1 sync.go:98] Error checking existing TLS certificate: error decoding cert PEM block
E1101 06:48:47.572767       1 sync.go:180] [kube-system/k8s-ui-init-ci-tls] Error getting certificate 'k8s-ui-init-ci-tls': error decoding cert PEM block
E1101 06:48:47.572910       1 runtime.go:66] Observed a panic: "index out of range" (runtime error: index out of range)
/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/panic.go:28
/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/sync.go:114
/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:236
/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:78
/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:199
/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:204
/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:168
/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:2197
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 78 [running]:
github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x126
panic(0x15fff40, 0x22054b0)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/jetstack-experimental/cert-manager/pkg/controller/certificates.(*Controller).Sync(0xc4203d0000, 0x21ba740, 0xc42031bb80, 0xc4206e4188, 0x21a3680, 0xc4202ac570)
	/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/sync.go:114 +0x9e4
github.com/jetstack-experimental/cert-manager/pkg/controller/certificates.(*Controller).processNextWorkItem(0xc4203d0000, 0x21ba740, 0xc42031bb80, 0xc4201e2480, 0x1e, 0x2, 0x2)
	/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:236 +0x2c5
github.com/jetstack-experimental/cert-manager/pkg/controller/certificates.(*Controller).(github.com/jetstack-experimental/cert-manager/pkg/controller/certificates.processNextWorkItem)-fm(0x21ba740, 0xc42031bb80, 0xc4201e2480, 0x1e, 0x2, 0x1476689)
	/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:78 +0x52
github.com/jetstack-experimental/cert-manager/pkg/controller/certificates.(*Controller).worker.func1(0xc4203d0000, 0xc4206f9d68, 0xc420368840, 0x1574ac0, 0xc4202c61e0, 0x0, 0x0)
	/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:199 +0x28f
github.com/jetstack-experimental/cert-manager/pkg/controller/certificates.(*Controller).worker(0xc4203d0000, 0xc420368840)
	/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:204 +0x1b5
github.com/jetstack-experimental/cert-manager/pkg/controller/certificates.(*Controller).Run.func1()
	/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:168 +0x33
github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4207d2880)
	/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x5e
github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4207d2880, 0x3b9aca00, 0x0, 0x1000000010001, 0xc420368840)
	/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4207d2880, 0x3b9aca00, 0xc420368840)
	/go/src/github.com/jetstack-experimental/cert-manager/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/jetstack-experimental/cert-manager/pkg/controller/certificates.(*Controller).Run
	/go/src/github.com/jetstack-experimental/cert-manager/pkg/controller/certificates/controller.go:168 +0x317
```

**What you expected to happen**:
Certificate is issued normally.

**How to reproduce it (as minimally and precisely as possible)**:
Create a `Certificate` without commonName or dnsNames

**Anything else we need to know?**:
The error is actually down in the Certificate parsing which panics when the Certificate is invalid: https://github.com/jetstack-experimental/cert-manager/blob/master/pkg/util/pki/csr.go#L13

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: v1.8.2
Server Version: v1.8.1+coreos.0
```

** Edited to properly describe the problem - which was not what I originally diagnosed **

--- a/pkg/controller/certificates/sync.go
+++ b/pkg/controller/certificates/sync.go
@@ -91,6 +91,15 @@ func (c *Controller) Sync(ctx context.Context, crt *v1alpha1.Certificate) (err e
 		return err
 	}
 
+	expectedCN, err := pki.CommonNameForCertificate(crt)
+	if err != nil {
+		return err
+	}
+	expectedDNSNames, err := pki.DNSNamesForCertificate(crt)
+	if err != nil {
+		return err
+	}
+
 	// grab existing certificate and validate private key
 	cert, err := kube.SecretTLSCert(c.secretLister, crt.Namespace, crt.Spec.SecretName)
 	if err != nil {
@@ -111,14 +120,6 @@ func (c *Controller) Sync(ctx context.Context, crt *v1alpha1.Certificate) (err e
 	defer c.scheduleRenewal(crt)
 
 	crtCopy := crt.DeepCopy()
-	expectedCN, err := pki.CommonNameForCertificate(crtCopy)
-	if err != nil {
-		return err
-	}
-	expectedDNSNames, err := pki.DNSNamesForCertificate(crtCopy)
-	if err != nil {
-		return err
-	}
 
 	// if the certificate was not found, or the certificate data is invalid, we
 	// should issue a new certificate.

--- a/pkg/controller/certificates/sync.go
+++ b/pkg/controller/certificates/sync.go
@@ -111,8 +111,14 @@ func (c *Controller) Sync(ctx context.Context, crt *v1alpha1.Certificate) (err e
 	defer c.scheduleRenewal(crt)
 
 	crtCopy := crt.DeepCopy()
-	expectedCN := pki.CommonNameForCertificate(crtCopy)
-	expectedDNSNames := pki.DNSNamesForCertificate(crtCopy)
+	expectedCN, err := pki.CommonNameForCertificate(crtCopy)
+	if err != nil {
+		return err
+	}
+	expectedDNSNames, err := pki.DNSNamesForCertificate(crtCopy)
+	if err != nil {
+		return err
+	}
 
 	// if the certificate was not found, or the certificate data is invalid, we
 	// should issue a new certificate.

--- a/pkg/issuer/acme/prepare.go
+++ b/pkg/issuer/acme/prepare.go
@@ -213,25 +213,12 @@ func authorizationsMap(list []v1alpha1.ACMEDomainAuthorization) map[string]v1alp
 	return out
 }
 
-func removeDuplicates(in []string) []string {
-	var found []string
-Outer:
-	for _, i := range in {
-		for _, i2 := range found {
-			if i2 == i {
-				continue Outer
-			}
-		}
-		found = append(found, i)
-	}
-	return found
-}
-
 func (a *Acme) authorizationsToObtain(ctx context.Context, cl *acme.Client, crt *v1alpha1.Certificate) ([]string, error) {
 	authMap := authorizationsMap(crt.Status.ACMEStatus().Authorizations)
-	expectedCN := pki.CommonNameForCertificate(crt)
-	expectedDNSNames := pki.DNSNamesForCertificate(crt)
-	check := removeDuplicates(append(expectedDNSNames, expectedCN))
+	expectedDNSNames, err := pki.DNSNamesForCertificate(crt)
+	if err != nil {
+		return nil, err
+	}
 	toAuthorize := util.StringFilter(func(domain string) (bool, error) {
 		auth, ok := authMap[domain]
 		glog.Infof("Compare %q with %q", auth.Account, a.issuer.GetStatus().ACMEStatus().URI)
@@ -239,7 +226,7 @@ func (a *Acme) authorizationsToObtain(ctx context.Context, cl *acme.Client, crt 
 			return false, nil
 		}
 		return checkAuthorization(ctx, cl, auth.URI)
-	}, check...)
+	}, expectedDNSNames...)
 
 	domains := make([]string, len(toAuthorize))
 	for i, v := range toAuthorize {

--- a/pkg/issuer/ca/issue.go
+++ b/pkg/issuer/ca/issue.go
@@ -121,7 +121,15 @@ func createCertificateTemplate(publicKey interface{}, commonName string, altName
 // publicKey is the public key of the signee, and signerKey is the private
 // key of the signer.
 func signCertificate(crt *v1alpha1.Certificate, issuerCert *x509.Certificate, publicKey interface{}, signerKey interface{}) ([]byte, *x509.Certificate, error) {
-	template, err := createCertificateTemplate(publicKey, pki.CommonNameForCertificate(crt), pki.DNSNamesForCertificate(crt)...)
+	cn, err := pki.CommonNameForCertificate(crt)
+	if err != nil {
+		return nil, nil, err
+	}
+	dnsNames, err := pki.DNSNamesForCertificate(crt)
+	if err != nil {
+		return nil, nil, err
+	}
+	template, err := createCertificateTemplate(publicKey, cn, dnsNames...)
 	if err != nil {
 		return nil, nil, fmt.Errorf("error creating x509 certificate template: %s", err.Error())
 	}

--- a/pkg/util/filter.go
+++ b/pkg/util/filter.go
@@ -57,15 +57,15 @@ func StringFilter(fn FilterFn, in ...string) StringFilterWrapper {
 }
 
 func RemoveDuplicates(in []string) []string {
-	dedupMap := make(map[string]struct{})
+	var found []string
+Outer:
 	for _, i := range in {
-		dedupMap[i] = struct{}{}
+		for _, i2 := range found {
+			if i2 == i {
+				continue Outer
+			}
+		}
+		found = append(found, i)
 	}
-	out := make([]string, len(dedupMap))
-	i := 0
-	for s := range dedupMap {
-		out[i] = s
-		i++
-	}
-	return out
+	return found
 }

--- a/pkg/util/filter.go
+++ b/pkg/util/filter.go
@@ -55,3 +55,17 @@ func StringFilter(fn FilterFn, in ...string) StringFilterWrapper {
 	}
 	return res
 }
+
+func RemoveDuplicates(in []string) []string {
+	dedupMap := make(map[string]struct{})
+	for _, i := range in {
+		dedupMap[i] = struct{}{}
+	}
+	out := make([]string, len(dedupMap))
+	i := 0
+	for s := range dedupMap {
+		out[i] = s
+		i++
+	}
+	return out
+}

--- a/pkg/util/filter_test.go
+++ b/pkg/util/filter_test.go
@@ -1,0 +1,36 @@
+package util
+
+import "testing"
+
+func TestRemoveDuplicates(t *testing.T) {
+	type testT struct {
+		input  []string
+		output []string
+	}
+	tests := []testT{
+		{
+			input:  []string{"a"},
+			output: []string{"a"},
+		},
+		{
+			input:  []string{"a", "b"},
+			output: []string{"a", "b"},
+		},
+		{
+			input:  []string{"a", "a"},
+			output: []string{"a"},
+		},
+		{
+			input:  []string{"a", "b", "a", "a", "c"},
+			output: []string{"a", "b", "c"},
+		},
+	}
+	for _, test := range tests {
+		actualOutput := RemoveDuplicates(test.input)
+		if len(actualOutput) != len(test.output) ||
+			!EqualUnsorted(test.output, actualOutput) {
+			t.Errorf("returned %q for %q but expected %q", actualOutput, test.input, test.output)
+			continue
+		}
+	}
+}

--- a/pkg/util/pki/csr.go
+++ b/pkg/util/pki/csr.go
@@ -3,28 +3,36 @@ package pki
 import (
 	"crypto/x509"
 	"crypto/x509/pkix"
+	"fmt"
 	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	"github.com/jetstack/cert-manager/pkg/util"
 )
 
 // The provided Certificate *must* specify either a DNS name or a
 // CommonName else this function will panic.
-func CommonNameForCertificate(crt *v1alpha1.Certificate) string {
-	if crt.Spec.CommonName == "" {
-		return crt.Spec.DNSNames[0]
+func CommonNameForCertificate(crt *v1alpha1.Certificate) (string, error) {
+	if crt.Spec.CommonName != "" {
+		return crt.Spec.CommonName, nil
 	}
-	return crt.Spec.CommonName
+	if len(crt.Spec.DNSNames) == 0 {
+		return "", fmt.Errorf("certificate must specify at least one of dnsNames or commonName")
+	}
+	return crt.Spec.DNSNames[0], nil
 }
 
 // The provided Certificate *must* specify either a DNS name or a
 // CommonName else this function will panic.
-func DNSNamesForCertificate(crt *v1alpha1.Certificate) []string {
+func DNSNamesForCertificate(crt *v1alpha1.Certificate) ([]string, error) {
 	if len(crt.Spec.DNSNames) == 0 {
-		return []string{crt.Spec.CommonName}
+		if crt.Spec.CommonName == "" {
+			return nil, fmt.Errorf("certificate must specify at least one of dnsNames or commonName")
+		}
+		return []string{crt.Spec.CommonName}, nil
 	}
 	if crt.Spec.CommonName != "" {
-		return append([]string{crt.Spec.CommonName}, crt.Spec.DNSNames...)
+		return util.RemoveDuplicates(append([]string{crt.Spec.CommonName}, crt.Spec.DNSNames...)), nil
 	}
-	return crt.Spec.DNSNames
+	return crt.Spec.DNSNames, nil
 }
 
 func GenerateCSR(commonName string, altNames ...string) *x509.CertificateRequest {

--- a/pkg/util/pki/csr_test.go
+++ b/pkg/util/pki/csr_test.go
@@ -1,0 +1,128 @@
+package pki
+
+import (
+	"testing"
+
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+)
+
+func buildCertificate(cn string, dnsNames ...string) *v1alpha1.Certificate {
+	return &v1alpha1.Certificate{
+		Spec: v1alpha1.CertificateSpec{
+			CommonName: cn,
+			DNSNames:   dnsNames,
+		},
+	}
+}
+
+func TestCommonNameForCertificate(t *testing.T) {
+	type testT struct {
+		name        string
+		crtCN       string
+		crtDNSNames []string
+		expectedCN  string
+		expectErr   bool
+	}
+	tests := []testT{
+		{
+			name:       "certificate with CommonName set",
+			crtCN:      "test",
+			expectedCN: "test",
+		},
+		{
+			name:        "certificate with one DNS name set",
+			crtDNSNames: []string{"dnsname"},
+			expectedCN:  "dnsname",
+		},
+		{
+			name:      "certificate with neither common name or dnsNames set",
+			expectErr: true,
+		},
+		{
+			name:        "certificate with both common name and dnsName set",
+			crtCN:       "cn",
+			crtDNSNames: []string{"dnsname"},
+			expectedCN:  "cn",
+		},
+		{
+			name:        "certificate with multiple dns names set",
+			crtDNSNames: []string{"dnsname1", "dnsname2"},
+			expectedCN:  "dnsname1",
+		},
+	}
+	testFn := func(test testT) func(*testing.T) {
+		return func(t *testing.T) {
+			actualCN, err := CommonNameForCertificate(buildCertificate(test.crtCN, test.crtDNSNames...))
+			if err != nil && !test.expectErr {
+				t.Errorf("did not expect error from CommonNameForCertificate: %s", err.Error())
+				return
+			}
+			if actualCN != test.expectedCN {
+				t.Errorf("expected %q but got %q", test.expectedCN, actualCN)
+				return
+			}
+		}
+	}
+	for _, test := range tests {
+		t.Run(test.name, testFn(test))
+	}
+}
+
+func TestDNSNamesForCertificate(t *testing.T) {
+	type testT struct {
+		name           string
+		crtCN          string
+		crtDNSNames    []string
+		expectDNSNames []string
+		expectErr      bool
+	}
+	tests := []testT{
+		{
+			name:           "certificate with CommonName set",
+			crtCN:          "test",
+			expectDNSNames: []string{"test"},
+		},
+		{
+			name:           "certificate with one DNS name set",
+			crtDNSNames:    []string{"dnsname"},
+			expectDNSNames: []string{"dnsname"},
+		},
+		{
+			name:      "certificate with neither common name or dnsNames set",
+			expectErr: true,
+		},
+		{
+			name:           "certificate with both common name and dnsName set",
+			crtCN:          "cn",
+			crtDNSNames:    []string{"dnsname"},
+			expectDNSNames: []string{"cn", "dnsname"},
+		},
+		{
+			name:           "certificate with multiple dns names set",
+			crtDNSNames:    []string{"dnsname1", "dnsname2"},
+			expectDNSNames: []string{"dnsname1", "dnsname2"},
+		},
+	}
+	testFn := func(test testT) func(*testing.T) {
+		return func(t *testing.T) {
+			actualDNSNames, err := DNSNamesForCertificate(buildCertificate(test.crtCN, test.crtDNSNames...))
+			if err != nil && !test.expectErr {
+				t.Errorf("did not expect error from CommonNameForCertificate: %s", err.Error())
+				return
+			}
+			if len(actualDNSNames) != len(test.expectDNSNames) {
+				t.Errorf("expected %q but got %q", test.expectDNSNames, actualDNSNames)
+				return
+			}
+			for i, actual := range actualDNSNames {
+				if test.expectDNSNames[i] != actual {
+					t.Errorf("expected %q but got %q", test.expectDNSNames, actualDNSNames)
+					return
+				}
+			}
+		}
+	}
+	for _, test := range tests {
+		t.Run(test.name, testFn(test))
+	}
+}

--- a/pkg/util/pki/csr_test.go
+++ b/pkg/util/pki/csr_test.go
@@ -102,6 +102,18 @@ func TestDNSNamesForCertificate(t *testing.T) {
 			crtDNSNames:    []string{"dnsname1", "dnsname2"},
 			expectDNSNames: []string{"dnsname1", "dnsname2"},
 		},
+		{
+			name:           "certificate with dnsName[0] set to equal common name",
+			crtCN:          "cn",
+			crtDNSNames:    []string{"cn", "dnsname"},
+			expectDNSNames: []string{"cn", "dnsname"},
+		},
+		{
+			name:           "certificate with a dnsName equal to cn",
+			crtCN:          "cn",
+			crtDNSNames:    []string{"dnsname", "cn"},
+			expectDNSNames: []string{"cn", "dnsname"},
+		},
 	}
 	testFn := func(test testT) func(*testing.T) {
 		return func(t *testing.T) {

